Wannier90 parser

JosePizarro3 opened the pull request:
Created a new parser for Wannier90 with the corresponding tests.

I also created new metainfo for Method, called `Projections`, and included the mainfile (`*.wout`) in the MatchingParser parsers.py file.

_Edit_: I indeed created a bit of a mess. Sorry for that :)

JosePizarro3 committed:
> Added Wannier90 parser

JosePizarro3 committed:
> Added Wannier90 tests

ladinesa commented on electronicparsers/wannier90/metainfo/wannier90.py:
```diff
@@ -0,0 +1,68 @@
+
+    m_def = Section(validate=False, extends_base_section=True)
+
+    x_wannier90_hoppings = SubSection(sub_section=x_wannier90_hopping_parameters, repeats=False)
```
shouldn't this be in method section?

JosePizarro3 replied:
x_wannier90_hoppings is the output, so I understand it should be in Calculation

ladinesa commented on electronicparsers/wannier90/parser.py:
```diff
@@ -0,0 +1,346 @@
+
+class Wannier90Parser:
+    # level defined w.r.t. DFT calculation (default, level = 0)
+    level = 1
```
I am actually not sure if this level is passed on to the MatchingParserInterface which is the nomad parser. I will check later

JosePizarro3 replied:
Ok, I can leave it like this then with a TODO comment

ladinesa commented on electronicparsers/wannier90/parser.py:
```diff
@@ -0,0 +1,346 @@
+        sec_wannier.hopping_matrix = np.array(np.array_split(full_hoppings, sec_wannier.nrpts))
+
+    def get_k_points(self):
+        reciprocal_lattice_vectors = np.vstack(self.wout_parser.get('reciprocal_lattice_vectors'))
```
assert first that self.wout_parser.get('reciprocal_lattice_vectors') is not None before stacking it

ladinesa commented on electronicparsers/wannier90/parser.py:
```diff
@@ -0,0 +1,346 @@
+        if reciprocal_lattice_vectors is None:
+            return
+
+        n_k_segments = self.wout_parser.get('n_k_segments')
```
put [] as defaut here. n_k_segments = self.wout_parser.get('n_k_segments', [])

ladinesa commented on electronicparsers/wannier90/parser.py:
```diff
@@ -0,0 +1,346 @@
+        sec_k_band = sec_scc.m_create(BandStructure, Calculation.band_structure_electronic)
+        sec_k_band.energy_fermi = energy_fermi
+
+        data = np.transpose(self.band_dat_parser.data)
```
make an assertion here that self.band_dat_parser.data is not empty.

ladinesa commented on tests/test_wannier90parser.py:
```diff
@@ -0,0 +1,82 @@
+def parser():
+    return Wannier90Parser()
+
+
```
maybe add a test or extend existing for other quantities like version number, system quantities...

ladinesa commented on tests/test_wannier90parser.py:
```diff
@@ -0,0 +1,82 @@
+        len(sec_scc[0].band_structure_electronic[0].segment[0].energies[0])
+    assert sec_scc[0].energy.fermi == sec_scc[0].band_structure_electronic[0].energy_fermi
+    assert sec_scc[0].band_structure_electronic[0].energy_fermi.to('eV').magnitude == 11.375
+
```
can you also assert the x_wannier_* quantities?

JosePizarro3 replied:
They asserted in the test_hoppings function.
Do you want me to put all the Calculation tests (for bands, dos, and x_wannier90) in the same function, e.g. `def test_scc`?

ladinesa replied:
It is only now that I have noticed that you only have one mainfile so it would make sense to put them all in one test

JosePizarro3 committed:
> Style + tests changes

JosePizarro3 committed:
> Fix english typo

JosePizarro3 committed:
> Energy fermi extracted from onsite energy for the orbital=1

> Fixed errors with regex

ladinesa commented on electronicparsers/wannier90/parser.py:
```diff
@@ -172,9 +172,9 @@ def parse_method(self):
         for key in self._input_projection_mapping.keys():
             setattr(sec_proj, self._input_projection_mapping[key], self.wout_parser.get(key))
         if self.wout_parser.get('Niter') is not None:
```
You can shorten this as `sec_proj.is_maximally_localized = self.wout_parser.get('Niter', 0) > 1`

ladinesa commented on tests/test_wannier90parser.py:
```diff
@@ -51,7 +51,7 @@ def test_lco(parser):
     assert sec_projection.n_projected_orbitals == 1
     assert sec_projection.n_bands == 5
-    assert sec_projection.is_maximally_localise is True
+    assert sec_projection.is_maximally_localized is True
```
This should be `assert sec_projection.is_maximally_localized`

JosePizarro3 committed:
> Fix pylint

JosePizarro3 commented:
@ladinesa I see now some errors with HoppingMatrix not being able to be imported.

No idea what's going on.

ladinesa commented:
This is normal as I have said before that the nomad pypi package does not yet include the metainfo defs for these.

JosePizarro3 commented:
Then the pylint is solved, let me know if I missed something else

ladinesa commented:
@JosePizarro3 please communicate with @ndaelman-hu about the merging since he is merging his stuffs with this branch. Or I suggest merge this one now then just change target branch for Nathan's merges. Please do the merge. And please do not forget to update nomad.

ndaelman-hu commented:
We are in the process of merging them atm, will soon be fully done.

JosePizarro3 commented:
Then let's wait until Nathan merges here, and then we merge wannier90_2 into the develop

ndaelman-hu committed:
Develop hubbard correction (#13)

* Add DFT+U processing and test to VASP parser

* Restructure data folder for VASP tests into matching subfolders

* Add DFT+U processing and test to FHI-aims parser

* Update base matching pattern (for text) to enable matching of repeated quantities on a line
Streamline mapping out of Hubbard keywords to `incar`

* Update spelling of the metadata to small caps

* Apply feedback from Pull request https://github.com/nomad-coe/electronic-parsers/pull/13

* Resolve pylint error for `nomad.datamodel.metainfo.simulation.method.HubbardModel`

* Update `fplo` metainfo in line with new `HubbardModel`
Revert solution for pylint: may be resolved with next pypi update

Co-authored-by: ndaelman <ndaelman@physik.hu-berlin.de>

ndaelman-hu committed:
Hybrid mixing coefficient  (#12)

* Direct hybrid alpha in the VASP parser to a standardized location in `run.method`

* Direct hybrid alpha in the FHI-aims parser to a standardized location in `run.method`

* Add FHI-aims data to previous commit

* Add hybrid name normalization to the parser + tests

Co-authored-by: ndaelman <ndaelman@physik.hu-berlin.de>

JosePizarro3 commented:
@ndaelman-hu @ladinesa should I then merge? 

I see some pylint errors still from HubbardModel not being recognized; I saw the [Develop hubbard correction](https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-FAIR/-/merge_requests/920) is closed without merging, but maybe this metainfo is somewhere else (?)

ndaelman-hu commented:
> I saw the [Develop hubbard correction](https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-FAIR/-/merge_requests/920) is closed without merging, but maybe this metainfo is somewhere else (?)
There were some merge conflicts, since `develop-hybrid-mixing` and `develop-hubbard-correction` both touch the code in the same place.
These conflicts have been solved and so, should soon be merged.

To my understanding though, what's holding back this pull request is the broken module, but @ladinesa can verify this better.

ladinesa commented:
You can merge this. It will pass later if nomad pypi is released with the updated metainfo.
